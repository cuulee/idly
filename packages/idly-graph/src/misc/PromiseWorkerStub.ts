import { GetActionTypes } from '../operations/operationsTypes';

// TOFIX WorkerGetStateActionsType only does GetState actions, want to add SetState aswell
export class PromiseWorkerStub {
  /* tslint:disable */
  public _workerCb: (message: GetActionTypes) => Promise<any>;
  registerPromiseWorker(workerCb: (message: GetActionTypes) => Promise<any>) {
    this._workerCb = workerCb;
  }
  /* tslint:enable */
  public postMessage(data: {
    readonly request: any;
    readonly type: string;
  }): Promise<any> {
    const t: GetActionTypes = JSON.parse(JSON.stringify(data));
    if (!t.type) {
      throw new Error('need type');
    }
    if (!t.request) {
      throw new Error('need request');
    }
    return this._workerCb(t);
  }
}
