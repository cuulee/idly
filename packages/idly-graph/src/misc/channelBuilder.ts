import { WorkerActionsType } from '../operations/types';

/**
 * A wrapper around `promiseWorker.postMessage` api
 * its sole purpose is to provide typings and easier testing
 * the main <-> worker calls.
 * @param promiseWorker
 */
export function channelBuilder<T extends WorkerActionsType>(promiseWorker: {
  readonly postMessage: any;
}): (type: T['type']) => (request: T['request']) => Promise<string> {
  return type => {
    return request => {
      const toSend = {
        request,
        type,
      };
      return promiseWorker.postMessage(toSend).catch((e: Error) => {
        console.log('Worker Error', e.message);
        return Promise.reject(e.message);
      });
    };
  };
}
