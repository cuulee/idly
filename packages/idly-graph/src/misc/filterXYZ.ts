import area from '@turf/area';
import bboxClip from '@turf/bbox-clip';
import bboxPolygon from '@turf/bbox-polygon';

import { BBox } from 'idly-common/lib/geo/bbox';
import { mercator } from 'idly-common/lib/geo/sphericalMercator';
import { Tile } from 'idly-common/lib/geo/tile';

export function filterXyz(xyzs: Tile[], bbox: BBox, ratio: number): Tile[] {
  const mainPoly = bboxPolygon(bbox);
  return xyzs.filter(tile => {
    const secPoly = bboxPolygon(mercator.bbox(tile.x, tile.y, tile.z));
    const inter = bboxClip(secPoly, bbox);
    return area(inter) / area(secPoly) >= ratio;
  });
}
