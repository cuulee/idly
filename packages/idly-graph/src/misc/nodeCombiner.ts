import { Feature, Point } from '@turf/helpers';
import { Node } from 'idly-common/lib/osm/structures';

export function nodeCombiner(node: Node, existingProps: {}): Feature<Point> {
  return {
    geometry: {
      coordinates: [node.loc.lon, node.loc.lat],
      type: 'Point',
    },
    id: node.id,
    properties: {
      ...existingProps,
      id: node.id,
    },
    type: 'Feature',
  };
}
