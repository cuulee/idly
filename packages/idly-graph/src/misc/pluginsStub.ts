import { EntityTable } from 'idly-common/lib/osm/structures';

export function pluginsStub(
  geometry: any = {
    'osm_basic--geometry': 'line',
  },
): Promise<any> {
  return Promise.resolve({
    workers: [
      {
        pluginName: 'dummy',
        // tslint:disable no-expression-statement
        worker: (entityTable: EntityTable) => {
          const m = new Map();
          entityTable.forEach((v, k) => {
            m.set(k, geometry);
          });
          return m;
        },
      },
    ],
  });
}
