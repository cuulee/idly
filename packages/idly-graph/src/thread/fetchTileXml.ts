import { mercator } from 'idly-common/lib/geo/sphericalMercator';
import { DOMParser } from 'xmldom';
import { BBox } from '@turf/helpers';

export async function fetchTileXml(
  x: number,
  y: number,
  zoom: number,
): Promise<string> {
  const xyz = [x, y, zoom].join(',');
  const bboxStr = mercator.bbox(x, y, zoom).join(',');
  const response = await fetch(
    `https://www.openstreetmap.org/api/0.6/map?bbox=${bboxStr}`,
  );
  if (!response.ok) {
    throw new Error(response.statusText);
  }
  const text = await response.text();
  return text;
}
