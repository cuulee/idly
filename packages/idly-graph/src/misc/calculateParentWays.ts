import { ImMap, ImSet } from 'idly-common/lib/misc/immutable';
import {
  EntityTable,
  EntityType,
  ParentWays,
} from 'idly-common/lib/osm/structures';

export function calculateParentWays(
  entityTable: EntityTable,
  parentWays: ParentWays = ImMap(),
): ParentWays {
  return parentWays.withMutations(p => {
    entityTable.forEach(w => {
      if (!w || w.type !== EntityType.WAY) {
        return;
      }
      w.nodes.forEach(nodeId => {
        /* tslint:disable */
        p.update(nodeId, (s = ImSet()) => s.add(w.id));
        /* tslint:enable */
      });
    });
  });
}
