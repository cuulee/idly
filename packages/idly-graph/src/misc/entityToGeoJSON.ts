import { Feature } from 'idly-common/lib/osm/feature';
import {
  Entity,
  EntityTable,
  EntityType,
} from 'idly-common/lib/osm/structures';

import { nodeCombiner } from './nodeCombiner';
import { wayCombiner } from './wayCombiner';

// tslint:disable:no-expression-statement object-literal-key-quotes
export function entityToGeoJson(
  entityTable: EntityTable,
  computedProps: any,
): Array<Feature<any, any>> {
  const arr: Array<Feature<any, any>> = [];
  entityTable.forEach((entity: Entity, id) => {
    if (entity.type === EntityType.NODE) {
      arr.push(nodeCombiner(entity, computedProps.get(id)));
    } else if (entity.type === EntityType.WAY) {
      arr.push(wayCombiner(entity, entityTable, computedProps.get(id)));
    } else if (entity.type === EntityType.RELATION) {
      // @TOFIX
    }
  });

  return arr;
}
