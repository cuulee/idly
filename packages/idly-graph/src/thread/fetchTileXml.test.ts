import { XML3 } from '../misc/fixtures';
import { fetchTileXml } from './fetchTileXml';

declare var global: any;

describe('fetchTileXml', () => {
  test('gives an xml', async () => {
    global.fetch = jest.fn().mockImplementation(() => {
      return new Promise((resolve, reject) => {
        resolve({
          id: '123',
          async text(): Promise<any> {
            return XML3;
          },
          ok: true,
        });
      });
    });
    const xml = await fetchTileXml(17, 17, 18);
    expect(xml).toMatchSnapshot();
  });
  test('should fail if status > 400', async () => {
    global.fetch = jest.fn().mockImplementation(() => {
      return new Promise((resolve, reject) => {
        resolve({
          id: '123',
          ok: false,
          statusText: 'errored',
          async text(): Promise<any> {
            return XML3;
          },
        });
      });
    });
    expect.assertions(1);
    fetchTileXml(17, 17, 18).catch(e => expect(e.message).toEqual('errored'));
  });
});
