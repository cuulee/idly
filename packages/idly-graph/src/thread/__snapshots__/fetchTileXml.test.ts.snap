// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchTileXml gives an xml 1`] = `
"
        <?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><osm>
        <node id=\\"1\\" visible=\\"true\\" version=\\"1\\" changeset=\\"38701677\\" timestamp=\\"2016-04-19T17:58:40Z\\" user=\\"Dombo\\" uid=\\"592097\\" lat=\\"34.5303960\\" lon=\\"74.1816382\\"/>
        <node id=\\"2\\" visible=\\"true\\" version=\\"1\\" changeset=\\"38701677\\" timestamp=\\"2016-04-19T17:58:40Z\\" user=\\"Dombo\\" uid=\\"592097\\" lat=\\"34.5303960\\" lon=\\"74.1816382\\"/>
        <node id=\\"3\\" visible=\\"true\\" version=\\"1\\" changeset=\\"38701677\\" timestamp=\\"2016-04-19T17:58:40Z\\" user=\\"Dombo\\" uid=\\"592097\\" lat=\\"34.5303960\\" lon=\\"74.1816382\\"/>
        <way id=\\"1\\" visible=\\"true\\" version=\\"2\\" changeset=\\"50089927\\" timestamp=\\"2017-07-06T17:19:04Z\\" user=\\"Srihari Thalla\\" uid=\\"2815653\\">
                <nd ref=\\"1\\"/>
                <nd ref=\\"2\\"/>
        </way>
 <way id=\\"2\\" visible=\\"true\\" version=\\"2\\" changeset=\\"50089927\\" timestamp=\\"2017-07-06T17:19:04Z\\" user=\\"Srihari Thalla\\" uid=\\"2815653\\">
                <nd ref=\\"1\\"/>
                <nd ref=\\"3\\"/>
        </way>
        </osm>"
`;
